Add matrix multiplication example (#5645)
STEllAR-GROUP/hpx: docs/sphinx/examples/matrix_multiplication.rst

..
    Copyright (C) 2021 Dimitra Karatza

    SPDX-License-Identifier: BSL-1.0
    Distributed under the Boost Software License, Version 1.0. (See accompanying
    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

> .. _examples_matrix_multiplication:

Reviewer: Somewhere in the example we should mention that this example is for demonstration purposes only, and that users should prefer to use optimized libraries like BLAS or blaze (https://bitbucket.org/blaze-lib/blaze, it even has an HPX backend!).

===================
Parallel algorithms
===================

This program will perform a matrix multiplication in parallel. The output will look something like this:

.. code-block:: text

   Matrix A is : 
   4 9 6 
   1 9 8 

   Matrix B is : 
   4 9 
   6 1 
   9 8 

   Resultant Matrix is : 
   124 93 
   111 127 

Setup
=====

The source code for this example can be found here:
:download:`matrix_multiplication.cpp
<../../examples/quickstart/matrix_multiplication.cpp>`.

To compile this program, go to your |hpx| build directory (see
:ref:`hpx_build_system` for information on configuring and building |hpx|) and
enter:

.. code-block:: shell-session

   $ make examples.quickstart.matrix_multiplication

To run the program type:

.. code-block:: shell-session

>    $ ./bin/matrix_multiplication

Reviewer: I think it'd be nice if this took the size of the matrices as command line options. The defaults can very well be 2x3 and 3x2 so you can leave the line above as it is, but if someone wants to play with the example it'd be nice to be able to change the sizes.

Defining three command-line options, `--n`, `--m`, `--k`, for the dimensions should be ok, where the first matrix is `n` x `m` and the second `m` x `k`.

This should print:

.. code-block:: text

   Matrix A is : 
   4 9 6 
   1 9 8 

   Matrix B is : 
   4 9 
   6 1 
   9 8 
   
   Resultant Matrix is : 
   124 93 
   111 127 

Notice that the numbers may be different because of the random initialization of the matrices.

Walkthrough
===========

Now that you have compiled and run the code, let's look at how the code works.

First, ``main()`` is used to initialize the runtime system and pass the command line arguments to the program. 
``hpx::init`` calls ``hpx_main()`` after setting up HPX, which is where our program is implemented.

.. literalinclude:: ../../examples/quickstart/matrix_multiplication.cpp
   :language: c++
   :start-after: //[mul_main
   :end-before: //]

> Going on with the ``hpx_main()`` function, we can see that matrix multiplication can be done very easily. 

Reviewer: ```suggestion
Continuing with the ``hpx_main()`` function, we can see that matrix multiplication can be done very easily. 
```

.. literalinclude:: ../../examples/quickstart/matrix_multiplication.cpp
   :language: c++
   :start-after: //[mul_hpx_main
   :end-before: //]

> First, the dimentions of the matrices is defined, so as that the number of columns of matrix A is equal 

Reviewer: ```suggestion
First, the dimensions of the matrices is defined, so as that the number of columns of matrix A is equal 
```

though this might need a small rewrite if the dimensions are given as command-line options.

to the number of rows of matrix B. You can freely change the dimentions as you wish. We use standard vectors 
to define the matrices to be multipled as well as the resultant matrix.

To give some random initial values to our matrices, we use `std::uniform_real_distribution
<https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution>`_. Then, ``hpx::generate()`` is used
along with ``std::bind()`` as an argument to yield two matrices A and B, which contain values in the range of [0, 10]. 

> You can also freely change the range of the values, if you wish.

Reviewer: Continuing with the command-line options, at least the seed should also be a command-line option, possibly also the range of values. You can seed the random number generator with something like this: https://github.com/STEllAR-GROUP/hpx/blob/dd3f3d9ae87b93c383877f4f69dc97842c2f8387/libs/core/async_cuda/tests/unit/cublas_matmul.cpp#L269-L273.

Author: what is `vm`?

Reviewer: `vm` is an object of type `variables_map`. Instead of `hpx_main` taking `argc/argv` you can have it take a `variables_map` instead. To make HPX aware of your custom command line options you have to tell it about them with an `options_description` like here: https://github.com/STEllAR-GROUP/hpx/blob/dd3f3d9ae87b93c383877f4f69dc97842c2f8387/libs/core/async_cuda/tests/unit/cublas_matmul.cpp#L330-L353.

> The next step is to perform the matriux multiplication in parallel. This can be done very easily by just using an 

Reviewer: ```suggestion
The next step is to perform the matrix multiplication in parallel. This can be done very easily by just using an 
```

Reviewer: ```suggestion
The next step is to perform the matriux multiplication in parallel. This can be done by just using an 
```
only my preference... I'd hope that the simplicity speaks for itself :P

.. code-block:: text

   Matrix A is : 
   4 9 6 
   1 9 8 

   Matrix B is : 
   4 9 
   6 1 
   9 8 
   
   Resultant Matrix is : 
   124 93 
   111 127 

Notice that the numbers may be different because of the random initialization of the matrices.

Walkthrough
===========

Now that you have compiled and run the code, let's look at how the code works.

First, ``main()`` is used to initialize the runtime system and pass the command line arguments to the program. 
``hpx::init`` calls ``hpx_main()`` after setting up HPX, which is where our program is implemented.

.. literalinclude:: ../../examples/quickstart/matrix_multiplication.cpp
   :language: c++
   :start-after: //[mul_main
   :end-before: //]

Proceeding to the ``hpx_main()`` function, we can see that matrix multiplication can be done very easily. 

.. literalinclude:: ../../examples/quickstart/matrix_multiplication.cpp
   :language: c++
   :start-after: //[mul_hpx_main
   :end-before: //]

First, the dimensions of the matrices are defined. If they were not given as command-line arguments, their default 
values are `2` x `3` for the first matrix and `3` x `2` for the second. We use standard vectors to define the matrices 

> to be multipled as well as the resultant matrix. 

Reviewer: ```suggestion
to be multiplied as well as the resultant matrix. 
```
as reported by our spell checker.

   :language: c++
   :start-after: //[mul_hpx_main
   :end-before: //]

First, the dimensions of the matrices are defined. If they were not given as command-line arguments, their default 
values are `2` x `3` for the first matrix and `3` x `2` for the second. We use standard vectors to define the matrices 
to be multipled as well as the resultant matrix. 

To give some random initial values to our matrices, we use `std::uniform_int_distribution
<https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution>`_. Then, ``std::bind()`` is used
along with ``hpx::ranges::generate()`` to yield two matrices A and B, which contain values in the range of [0, 10] or in
the range defined by the user at the command-line arguments. The seed to generate the values can also be defined by the user.

The next step is to perform the matrix multiplication in parallel. This can be done by just using an ``hpx::for_loop()`` 
combined with a parallel execution policy ``hpx::execution::par`` as the outer loop of the multiplication. Note that the execution 
of ``hpx::for_loop()`` without specifying an execution policy is equivalent to specifying hpx::execution::seq 

> as the execution policy. More information about ``hpx::for_loop()`` can be found in :cpp:func:`\hpx::for_loop`.

Reviewer: Do you think it would make more sense to just make all the mentions of `hpx::for_loop` links to the API reference (i.e. `:cpp:func:` links)? Like this it's going to render as "More information about `hpx::for_loop` can be found in `hpx::for_loop`" with the latter being a link, which to me looks a bit funny and redundant.
